Converted CSS to SASS

heartsucker opened the pull request:
I wanted to a add new features and CSS and is terrible to work with compared to SASS. This PR includes:
- Convert all CSS to SASS
- Combine duplicate SASS rules
- Add SASS compilation to Ansible / builds

garrettr commented:
Thanks for this PR, @heartsucker! I've been wanting to do this for a long time, since I agree CSS is terrible to work with compared to alternatives such as SASS. I'll review this by the end of the week.

heartsucker commented:
Hey guys. This is in need of some rebasing, but before I get to it, can someone take a look and let me know if this seems to be a good candidate for merge?

garrettr commented on securedrop/sass/source.sass:
```diff
@@ -0,0 +1,206 @@
+     * HACK: Firefox likes to add non-existent leading and/or trailing whitespace
+     * to <span>s, but never does it to <p>s. This hack uses a <p> for the
+     * codename so we can get the right copy behavior, and uses inline-block plus
+     * tweaks the margin to make it look like a span
```
Close this comment.

garrettr commented on securedrop/sass/source.sass:
```diff
@@ -0,0 +1,206 @@
+        font-size: 14px
+
+        p
+            *
```
Is this line necessary?

heartsucker replied:
Probably not? I'm not a CSS expert, so I made as few changes as possible when porting it to SASS.

garrettr replied:
@heartsucker Gotcha. Let's leave it in for now, it can be addressed in a follow-up if necessary.

garrettr commented on securedrop/source_templates/base.html:
```diff
@@ -3,8 +3,9 @@
-    {% assets filters="cssmin", output="gen/source.css",
-      "css/normalize.css", "css/source.css", "css/font-awesome.css" %}
+
+    {# TODO why doesn't a link to /static/source.css work? #}
```
We should resolve this TODO before merging. It's best not to leave TODOs in merged code, if possible.

heartsucker replied:
I'll take a look at this since there have been some updates to some of the dependencies since I wrote this and the behavior might have changed. Also, suggestions welcome since this one seems a bit non-intuitive.

garrettr commented on install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/tasks/build_securedrop_app_code_deb.yml:
```diff
@@ -48,6 +57,15 @@
+  args:
+    chdir: /tmp/{{ securedrop_app_code_deb }}/var/www/securedrop/
+#  register: sass_compile_result
+#  changed_when: sass_compile_result.stdout != ''
```
This code is commented out, can it be removed?

heartsucker commented on securedrop/management/run.py:
```diff
@@ -151,12 +154,15 @@ def run():
+                                 'cyan'),
+        lambda: DevServerProcess('SASS Compiler',
+                                 ['sass', '--watch', 'sass:static/css'],
+                                 'magenta'),
```
If one of the subprocesses fails to start or throws and exception, the monitor will leave the other processes dangling. Each call to DevServerProccess needs a try/except around it.

garrettr replied:
@heartsucker Thanks for the explanation! I'll take a look at this myself - I think there might be a way to more gracefully handle that case within DevServerProcess.

heartsucker commented:
@garrettr I'm not really a web frontend guy, but more a guy who can do some frontend stuff when he needs. I happened to do it in SASS because that's what I know, but some poking around does seem to indicate that there are more advantages to SCSS. If y'all want that, I can make those change along with all of the rest of this.

fowlslegs commented:
> I'm not really a web frontend guy

Me either.

> poking around does seem to indicate that there are more advantages to SCSS.

From a little research, I saw a decent amount of opinion to that point as well. E.g., http://thesassway.com/editorial/sass-vs-scss-which-syntax-is-better.

>  If y'all want that, I can make those change along with all of the rest of this.

If that isn't too much work for you, it seems to me like that might be good, just because it seems like SCSS might be better for other contributors to work with in the future. Since I don't do much myself with the frontend codebase myself, I'm not going to make an absolute call one way or the other. Unless one of the other SD team members has strong opinions, we'll leave the decision up to you.

garrettr commented:
> I happened to do it in SASS because that's what I know, but some poking around does seem to indicate that there are more advantages to SCSS. If y'all want that, I can make those change along with all of the rest of this.

@heartsucker I think SASS is fine. I polled some front-end folks that I know and people seem to be pretty evenly split on which one is "better". If we end up deciding to change to SCSS later, that seems to be trivial to do via `sass-convert`.

heartsucker commented:
This was tested on the dev VM, but it's not passing Travis because of this mystery:

```
TASK: [development | Install sass Ruby gem] *********************************** 
<localhost> state=latest name=sass
changed: [localhost] => {"changed": true, "name": "sass", "state": "latest", "version": "3.4.22"}

TASK: [development | Compile sass to css] ************************************* 
<localhost> chdir=/home/travis/build/freedomofpress/securedrop/securedrop
failed: [localhost] => {"changed": true, "cmd": "sass --update sass:static/css", "delta": "0:00:00.004013", "end": "2016-12-03 15:14:05.668050", "rc": 127, "start": "2016-12-03 15:14:05.664037", "warnings": []}
stderr: /bin/sh: 1: sass: not found
```

It's saying the gem was installed (globally), but then it's not on the path. Anyone have any experience debugging the Travis environment? Also, I added verbosity to build in hopes that it might tell us something, but alas.

redshiftzero commented:
Hey @heartsucker, I took a peek at this PR yesterday and it does look to be a weird PATH issue: one way to troubleshoot is to add a couple of post-install tasks in that Ansible play before you try to execute `sass`, e.g.:

```
- name: Check sass is in gem list
  command: gem list

- name: Check where gems are being installed
  command: gem environment
```

Since you put those -vv verbosity flags (good call) you can inspect the output and see what's going on with the PATH. This might not be the Best Way to debug this but it’s what I do  ¯\_(ツ)_/¯. Give that a whirl, if things still look confusing let me know and I’ll take a deeper look.

heartsucker commented:
@redshiftzero I already did most of that (and them some), but even when putting calls `sass` with the full path `/usr/lib/gems/sass` (or whatever it was), it still fails. I've never dealt with Ruby, so whatever seems to be going on with RVM is new to me. Anywho. I can probably get to this this weekend.

heartsucker commented:
So I've spent a couple of hours on this, and all I have so show is a new error message:

```
/usr/bin/env: ruby_executable_hooks1.9.1: No such file or directory
```

Any rubyists have any idea what do to here?

heartsucker commented:
Ok, so in commit b9c78be I did just about the kludgiest thing I've done in a while. It works, but it makes the build take over a minute longer than it should.

This whole bit needs to happen because the SASS files need to be compiled to CSS since some of the functional tests require DOM elements to be visible.

heartsucker commented:
Pinging for review/merge.

heartsucker commented:
Doubling down on the request for merge because there are a bunch of UI tickets that got opened today, and it would be nice to do this first.

heartsucker commented:
@garrettr? @fowlslegs? @redshiftzero? Bueller?

conorsch commented:
Whoa, sorry about the lag on this one, @heartsucker! Thanks for your persistence, getting this in pronto. (Yes, the hack in b9c78be is indeed adventurous approach, but it works well enough for the dev env!)
